SEP-1932: DPoP Profile for MCP (#1932)
modelcontextprotocol/modelcontextprotocol: seps/0000-dpop-extension.md

# SEP-1932: DPoP Profile for MCP

> **Note**: This SEP defines an optional security extension for MCP that enables sender-constrained access tokens through DPoP (Demonstrating Proof-of-Possession).

- **Status**: Draft
- **Type**: Standards Track
- **Created**: 2025-12-05
- **Author(s)**: Pieter Kasselman
- **Sponsor**: Darin McAdams
- **PR**: https://github.com/modelcontextprotocol/ext-auth/blob/pieterkas-dpop-extension/specification/draft/dpop-extension.mdx

## Abstract

> This SEP defines an optional DPoP (Demonstrating Proof-of-Possession) extension for the Model Context Protocol that enables sender-constrained access tokens to prevent token replay attacks. The extension binds OAuth 2.0 access tokens to cryptographic key pairs controlled by MCP clients, requiring clients to prove possession of the private key with each request. A key enhancement specific to MCP's architecture is the required inclusion of a content digest in DPoP proofs, which cryptographically binds each proof to the specific JSON-RPC request body, addressing security challenges posed by MCP's single-endpoint architecture.

Reviewer: ```suggestion
This SEP defines an optional DPoP (Demonstrating Proof of Possession) extension for the Model Context Protocol that enables sender-constrained access tokens to prevent token replay attacks. The extension binds OAuth 2.0 access tokens to cryptographic key pairs controlled by MCP clients, requiring clients to prove possession of the private key with each request. A key enhancement specific to MCP's architecture is the required inclusion of a content digest in DPoP proofs, which cryptographically binds each proof to the specific JSON-RPC request body, addressing security challenges posed by MCP's single-endpoint architecture.
```

DPoP addresses this vulnerability by making tokens "sender-constrained": even if an attacker intercepts both the access token and a DPoP proof, they cannot use them without also possessing the client's private key. This significantly raises the bar for attackers.

MCP's architecture presents a unique security challenge: all JSON-RPC requests use the same HTTP endpoint and method (POST). Standard DPoP binds proofs to the endpoint and HTTP method but not to the request payload, allowing intercepted proofs to be replayed with different message bodies within the validity window. This SEP addresses this by requiring a content digest in DPoP proofs, binding each proof cryptographically to a specific request body.

This extension is particularly valuable for:

- High-security environments handling sensitive data
- Long-lived access tokens
- Deployments where token theft risk is elevated
- Compliance requirements mandating proof-of-possession

## Specification

> The DPoP Profile for MCP strengthens OAuth 2.0 sender-constrained tokens in the Model Context Protocol by addressing a security gap created by MCP’s single-endpoint architecture. Standard DPoP binds proofs to the HTTP method and URL, but in MCP all requests share the same method (POST) and endpoint, and the actual operation is encoded entirely in the JSON-RPC body. This makes traditional DPoP vulnerable to replay attacks where an intercepted proof could be reused with a different request payload.

Reviewer: This suggestion attempts to rephrase things in a more neutral and fact based tone.  Calling it a "security gap" and "vulnerable" is IMHO a bit hyperbolic. 

```suggestion
The DPoP Profile for MCP adapts OAuth 2.0 sender-constrained tokens for use in the Model Context Protocol’s single-endpoint architecture. Standard DPoP binds proofs to the HTTP method and URL, but in MCP all requests use the same method (POST) and endpoint, with the specific operation conveyed in the JSON-RPC body. As a result, traditional DPoP does not distinguish between different MCP operations, since the request semantics are not reflected in the elements covered by the proof. The profile refines the content of the DPoP proof to better account for this characteristic of MCP.
```

### 1. Single-Endpoint Architecture Increases Replay Risk

In typical RESTful APIs, DPoP proofs are inherently bound to the specific method and path of each request, limiting their replay value. MCP’s design, however, uses:

- A single HTTP endpoint, and
- A single HTTP method (`POST`),
- With all semantic variation encoded in the JSON-RPC message body.

This means the `htu` and `htm` claims in a standard DPoP proof carry little entropy—they do not meaningfully differentiate purposes or operations. As a result, an attacker who obtains a proof (and valid token) during the proof validity window could replay it with a different JSON-RPC payload, gaining unauthorized access to operations the original client never invoked.

> This is a threat unique to protocols like MCP that multiplex operations within a uniform transport envelope.

Reviewer: ```suggestion
This is unique to protocols like MCP that multiplex operations within a uniform transport envelope.
```

## Testing Plan

Implementations should cover:

1. **Proof generation**: Correct formatting of all required claims including content digest
2. **Content digest calculation**: Accurate SHA-256 hashing and base64 encoding
3. **Validation logic**: All validation steps per specification
4. **Error handling**: Appropriate HTTP 401 responses with error details
5. **Time window**: Correct handling of `iat` claims at boundaries
6. **Algorithm support**: All declared algorithms work correctly

Test vectors will be provided in the reference implementation.

## Alternatives Considered

### HTTP Message Signatures

We considered RFC 9421 (HTTP Message Signatures) as an alternative to content digests. However, this would add complexity without significant benefit for MCP's use case. Content digests provide the needed payload binding with simpler implementation requirements.

### Mandatory jti Tracking

Making `jti` tracking mandatory would provide stronger replay protection but at significant operational cost. The stateless design better fits MCP's scalability goals while still providing substantial security improvements through content digest validation.

### Nonce-based Approach

> Server-provided nonces (per RFC 9449 Section 9) could replace time-based validation but require additional round-trips and server state. This specification makes nonces optional, allowing implementations to choose based on their security requirements.

Reviewer: ```suggestion
Server-provided nonces (per RFC 9449 Section 9) could replace time-based validation but require additional round-trips and potentially server state. This specification makes nonces optional, allowing implementations to choose based on their security requirements.
```
